src/network/serverpackethandler.cpp

```diff
@@ -1187,11 +1187,17 @@ void Server::handleCommand_Interact(NetworkPacket *pkt)
 			// Get diggability and expected digging time
 			DigParams params = getDigParams(m_nodedef->get(n).groups,
 					&selected_item.getToolCapabilities(m_itemdef));
-			// If can't dig, try hand
-			if (!params.diggable) {
-				params = getDigParams(m_nodedef->get(n).groups,
+
+			// Get the digging parameters for the hand.
+			DigParams hand_params = getDigParams(m_nodedef->get(n).groups,
 					&hand_item.getToolCapabilities(m_itemdef));
+
+			// If the tool can't dig or takes more time than the hand,
+			// use the hand instead.
+			if(!params.diggable || (hand_params.diggable && params.time >= hand_params.time)) {
```

Comment:
Please adjust the cheat detection code accordingly, for cases where the digging times makes a big difference:
https://github.com/minetest/minetest/blob/dc165fe942bcc48d7dea0a7b722886937d9c6914/src/network/serverpackethandler.cpp#L1205-L1211

Reply:
I don't know where any changes would be necessary.
The new hand digging times only occur when the legitimate hand can dig faster. If the client is older (or tries to trick the server for some reason) and still uses the tool digging times, it just seems like the digging time is *longer* than expected, which doesn't seem to cause problems from by observations and tests (though I certainly don't understand the anticheat system entirely).

A newer client on an older server could run into trouble, but that would require backporting or another solution which may need to be addressed.

Reply:
Ah yes right.  Sorry, I don't know what I thought there. The code and what you wrote totally makes sense.